File: src/main/java/net/zhuruoling/omms/central/network/chatbridge/UdpBroadcastSender.java
package net.zhuruoling.omms.central.network.chatbridge;

import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashMap;
import java.util.concurrent.ConcurrentHashMap;

public class UdpBroadcastSender extends Thread {

    private final Logger logger = LoggerFactory.getLogger("UdpBroadcastSender");
    boolean stopped = false;
    private final ConcurrentHashMap<Target, byte[]> queue = new ConcurrentHashMap<>();
    private final HashMap<Target, MulticastSocket> multicastSocketCache = new HashMap<>();

    public UdpBroadcastSender() {
        this.setName("UdpBroadcastSender#" + this.getId());
    }

    public void createMulticastSocketCache(@NotNull Target target) {
        try {
            multicastSocketCache.put(target, createMulticastSocket(target.getAddress(), target.getPort()));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public void clearMulticastSocketCache(){
        multicastSocketCache.clear();
    }

    private static @NotNull MulticastSocket createMulticastSocket(String addr, int port) throws IOException {
        MulticastSocket socket;
        socket = new MulticastSocket(port);
        //socket.joinGroup(new InetSocketAddress(inetAddress, port), NetworkInterface.getByInetAddress(inetAddress));
        return socket;

    }

    @Override
    public void run() {
        logger.info("Starting UdpBroadcastSender.");
        while (!stopped) {
            try {
                if (!queue.isEmpty()) {
                    queue.forEach(this::send);
                }
                sleep(10);
            }catch (Exception e){
                e.printStackTrace();
            }
        }
        logger.info("Stopped!");
    }

    public boolean isStopped() {
        return stopped;
    }

    public void setStopped(boolean stopped) {
        this.stopped = stopped;
    }

    public void addToQueue(@NotNull Target target, @NotNull String content) {
        queue.put(target, content.getBytes(StandardCharsets.UTF_8));
    }

    private void send(@NotNull Target target, byte @NotNull [] content) {
        queue.remove(target,content);
        MulticastSocket socket;
        try {
            if (multicastSocketCache.containsKey(target)) {
                socket = multicastSocketCache.get(target);
            } else {
                socket = createMulticastSocket(target.getAddress(), target.getPort());
                multicastSocketCache.put(target,socket);
            }
            DatagramPacket packet = new DatagramPacket(content, content.length, InetAddress.getByName(target.getAddress()), target.getPort());

            socket.send(packet);
        } catch (Exception e) {
            logger.error("Cannot send UDP Broadcast.\n\tTarget=%s\n\tContent=%s"
                            .formatted(target.toString(), Arrays.toString(content))
                    , e);
        }
    }



}

File: src/main/java/icu/takeneko/omms/central/network/chatbridge/UdpBroadcastSender.kt
package icu.takeneko.omms.central.network.chatbridge

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.*
import java.nio.charset.StandardCharsets
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.locks.LockSupport

class UdpBroadcastSender : Thread() {
    private val logger: Logger = LoggerFactory.getLogger("UdpBroadcastSender")
    var isStopped: Boolean = false
    private val queue = LinkedBlockingQueue<Pair<UdpBroadcastTarget, ByteArray>>()
    private val multicastSocketCache = mutableMapOf<UdpBroadcastTarget, MulticastSocket>()

    init {
        this.name = "UdpBroadcastSender#" + this.id
    }

    fun createMulticastSocketCache(target: UdpBroadcastTarget) {
        try {
            multicastSocketCache[target] = createMulticastSocket(target.address, target.port)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    fun clearMulticastSocketCache() {
        multicastSocketCache.clear()
    }

    override fun run() {
        logger.info("Starting UdpBroadcastSender.")
        while (!isStopped) {
            try {
                while (queue.isNotEmpty()) {
                    val (t, b) = queue.poll()
                    this.send(t, b)
                }
                LockSupport.parkNanos(1000)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        logger.info("Stopping!")
    }

    fun addToQueue(target: UdpBroadcastTarget, content: String) {
        queue.add(target to content.toByteArray(StandardCharsets.UTF_8))
    }

    private fun send(target: UdpBroadcastTarget, content: ByteArray) {
        try {
            val socket = if (multicastSocketCache.containsKey(target)) {
                multicastSocketCache[target]!!
            } else {
                createMulticastSocket(target.address, target.port).also { multicastSocketCache[target] = it }
            }
            val packet = DatagramPacket(content, content.size, InetAddress.getByName(target.address), target.port)
            socket.send(packet)
        } catch (e: Exception) {
            logger.error("Send UDP Broadcast failed, target={}, content={}", target, content)
        }
    }

    private fun createMulticastSocket(addr: String, port: Int): MulticastSocket {
        val socket = MulticastSocket(port)
        val inetAddress = InetAddress.getByName(addr)
        socket.joinGroup(InetSocketAddress(addr, port), NetworkInterface.getByInetAddress(inetAddress))
        return socket
    }
}